### PR TITLE
Add WSD scheduler

### PR DESCRIPTION
**Pull request by visheratin**
# What does this PR do?

This PR adds the Warmup-Stable-Decay learning rate scheduler. The scheduler and its benefits are described in detail in the MiniCPM [blog post](https://shengdinghu.notion.site/MiniCPM-Unveiling-the-Potential-of-End-side-Large-Language-Models-d4d3a8c426424654a4e80e42a711cb20). In short, it allows continuous training and performs better than standard cosine scheduler.

Here is an example of learning rate dynamics with this schedule:
image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr, @pacman100

### CODE REVIEW

**Review by muellerzr — COMMENTED** [review]
Nice! Seems handy. Could you rebase from `main` and see if that fixes the failing tests?

**Review by muellerzr — APPROVED** [review]
LG2M, no idea why that pytorch test is failing 😕 

cc @amyeroberts for final review (and ideas(?) was seeing that on another PR of mine iirc)

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this!

**Comment by muellerzr** [comment]
@visheratin CI should be green on rebase, we found the issue :)
